File: Sources/Controllers/Misc/Views/ModalSheetVC.m
#import "ModalSheetVC.h"


@implementation ModalSheetVC {

	ModalChildView *modalChildView;
	PinCodeVC *pinCodeVC;
	UINavigationController *navVC;

}

// ! Lifecycle

- (id)init {

	self = [super init];
	if(!self) return nil;

	pinCodeVC = [PinCodeVC new];
	pinCodeVC.delegate = self;

	[self setupViews];

	return self;

}


- (void)viewDidAppear:(BOOL)animated {

	[super viewDidAppear:animated];
	[modalChildView animateViews];

}


- (void)setupViews {

	self.view.backgroundColor = UIColor.clearColor;

	modalChildView = [ModalChildView new];
	modalChildView.delegate = self;
	[self.view addSubview: modalChildView];

	[self layoutUI];

}

// ! Designated initializers

- (void)setupChildWithTitle:(NSString *)title
	subtitle:(NSString *)subtitle
	buttonTitle:(NSString *)buttonTitle
	forTarget:(id)target
	forSelector:(SEL)selector
	secondButtonTitle:(NSString *)secondTitle
	forTarget:(id)secondTarget
	forSelector:(SEL)secondSelector
	thirdStackView:(BOOL)thirdSV
	thirdButtonTitle:(NSString *)thirdTitle
	forTarget:(id)thirdTarget
	forSelector:(SEL)thirdSelector
	accessoryImage:(UIImage *)accessoryImage
	secondAccessoryImage:(UIImage *)secondAccessoryImg
	thirdAccessoryImage:(UIImage *)thirdAccessoryImg
	prepareForReuse:(BOOL)prepare
	scaleAnimation:(BOOL)scaleAnim {

	[modalChildView setupModalChildWithTitle:title
		subtitle:subtitle
		buttonTitle:buttonTitle
		forTarget:target
		forSelector:selector
		secondButtonTitle:secondTitle
		forTarget:secondTarget
		forSelector:secondSelector
		thirdStackView:thirdSV
		thirdButtonTitle:thirdTitle
		forTarget:thirdTarget
		forSelector:thirdSelector
		accessoryImage:accessoryImage
		secondAccessoryImage:secondAccessoryImg
		thirdAccessoryImage:thirdAccessoryImg
		prepareForReuse:prepare
		scaleAnimation:scaleAnim
	];

}


- (void)setupChildWithTitle:(NSString *)title
	subtitle:(NSString *)subtitle
	buttonTitle:(NSString *)buttonTitle
	forTarget:(id)target
	forSelector:(SEL)selector
	secondButtonTitle:(NSString *)secondTitle
	forTarget:(id)secondTarget
	forSelector:(SEL)secondSelector
	accessoryImage:(UIImage *)accessoryImage
	secondAccessoryImage:(UIImage *)secondAccessoryImg
	prepareForReuse:(BOOL)prepare
	scaleAnimation:(BOOL)scaleAnim {

	[modalChildView setupModalChildWithTitle:title
		subtitle:subtitle
		buttonTitle:buttonTitle
		forTarget:target
		forSelector:selector
		secondButtonTitle:secondTitle
		forTarget:secondTarget
		forSelector:secondSelector
		thirdStackView:NO
		thirdButtonTitle:nil
		forTarget:nil
		forSelector:nil
		accessoryImage:accessoryImage
		secondAccessoryImage:secondAccessoryImg
		thirdAccessoryImage:nil
		prepareForReuse:prepare
		scaleAnimation:scaleAnim
	];

}


- (void)shouldCrossDissolveChildSubviews { [modalChildView shouldCrossDissolveSubviews]; }
- (void)layoutUI { [self.view pinViewToAllEdges: modalChildView]; }
- (void)vcNeedsDismissal {

	[modalChildView animateDismissWithCompletion:^(BOOL finished) {
		[self dismissViewControllerAnimated:YES completion:nil];
	}];

}

// ! Reusable funcs

- (void)configureVC:(UIViewController *)vc
	withTitle:(NSString *)title
	withItemImage:(UIImage *)image
	forSelector:(SEL)selector
	isLeftBarButtonItem:(BOOL)leftItem {

	navVC = [[UINavigationController alloc] initWithRootViewController: vc];
	vc.title = title;
	if(leftItem) vc.navigationItem.leftBarButtonItem = [UIBarButtonItem
		getBarButtonItemWithImage:image
		forTarget:self
		forSelector:selector
	];
	else vc.navigationItem.rightBarButtonItem = [UIBarButtonItem
		getBarButtonItemWithImage:image
		forTarget:self
		forSelector:selector
	];
	navVC.modalTransitionStyle = UIModalTransitionStyleCrossDissolve;
	navVC.modalPresentationStyle = UIModalPresentationFullScreen;

}

- (void)dismissVC {

	[self dismissViewControllerAnimated:YES completion:nil];
	dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.8 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
		[modalChildView animateDismissWithCompletion:^(BOOL finished) {
			[self dismissViewControllerAnimated:YES completion:nil];
		}];
	});

}

// ! ModalChildViewDelegate

- (void)modalChildViewDidTapScanQRCodeButton {

	QRCodeVC *qrCodeVC = [QRCodeVC new];
	qrCodeVC.delegate = self;
	[self configureVC:qrCodeVC
		withTitle:@"Scan QR Code"
		withItemImage:[UIImage systemImageNamed:@"xmark.circle.fill"]
		forSelector:@selector(didTapDismissButton)
		isLeftBarButtonItem:YES
	];
	[self presentViewController:navVC animated:YES completion:nil];

}


- (void)modalChildViewDidTapImportQRImageButton {

	UIImagePickerController *pickerC = [UIImagePickerController new];
	pickerC.delegate = self;
	pickerC.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
	[self presentViewController:pickerC animated:YES completion:nil];

}


- (void)modalChildViewDidTapEnterManuallyButton {

	[self configureVC:pinCodeVC
		withTitle:@"Enter QR Code"
		withItemImage:[UIImage systemImageNamed: @"checkmark.circle.fill"]
		forSelector:@selector(didTapComposeButton)
		isLeftBarButtonItem:NO
	];
	pinCodeVC.navigationItem.leftBarButtonItem = [UIBarButtonItem
		getBarButtonItemWithImage:[UIImage systemImageNamed:@"xmark.circle.fill"]
		forTarget:self
		forSelector:@selector(didTapDismissButton)
	];
	[self presentViewController:navVC animated:YES completion:nil];

}


- (void)modalChildViewDidTapDimmedView {

	[modalChildView animateDismissWithCompletion:^(BOOL finished) {

		[self dismissViewControllerAnimated:YES completion:nil];

	}];

}


- (void)modalChildViewDidPanWithGesture:(UIPanGestureRecognizer *)panRecognizer
	modifyingConstraint:(NSLayoutConstraint *)constraint {

	CGPoint translation = [panRecognizer translationInView: self.view];
	CGFloat newHeight = modalChildView.currentSheetHeight - translation.y;

	switch(panRecognizer.state) {

		case UIGestureRecognizerStateChanged:

			if(newHeight < modalChildView.kDefaultHeight) {
				constraint.constant = newHeight;
				constraint.active = YES;

				[modalChildView calculateAlphaBasedOnTranslation: translation];

				[self.view layoutIfNeeded];
			}
			break;

		case UIGestureRecognizerStateEnded:

			if(newHeight < modalChildView.kDismissableHeight) [modalChildView animateDismissWithCompletion:^(BOOL finished) {
				[self dismissVC];
			}];
			else if(newHeight < modalChildView.kDefaultHeight) [modalChildView animateSheetHeight: modalChildView.kDefaultHeight];
			break;

		default: break;

	}

}

// ! PinCodeVCDelegate

- (void)pinCodeVCShouldDismissVC {

	[self.delegate modalSheetVCShouldReloadData];
	[self dismissVC];

}


- (void)pinCodeVCShouldPushAlgorithmVC {

	AlgorithmVC *algorithmVC = [AlgorithmVC new];
	algorithmVC.title = @"Algorithm";
	[navVC pushViewController:algorithmVC animated:YES];

}

// ! UIImagePickerControllerDelegate

- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary *)info {

	UIImage *chosenImage = info[UIImagePickerControllerOriginalImage];

	CIImage *ciImage = [[CIImage alloc] initWithImage: chosenImage];
	NSDictionary *options = @{ CIDetectorAccuracy: CIDetectorAccuracyHigh };
	CIDetector *qrDetector = [CIDetector detectorOfType:CIDetectorTypeQRCode context:nil options:options];

	NSArray *features = [qrDetector featuresInImage:ciImage options:options];
	if(features == nil || features.count == 0) return;

	for(CIQRCodeFeature *qrCodeFeature in features) {
		[[TOTPManager sharedInstance] makeURLOutOfOtPauthString: qrCodeFeature.messageString];
		[self.delegate modalSheetVCShouldReloadData];
	}

	[self dismissVC];

}

// ! QRCodeVCDelegate

- (void)qrCodeVCDidCreateIssuerOutOfQRCode {

	[self.delegate modalSheetVCShouldReloadData];
	[self dismissVC];

}


- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker { [self dismissVC]; }

// ! Selectors

- (void)didTapComposeButton {

	[NSNotificationCenter.defaultCenter postNotificationName:@"checkIfDataShouldBeSaved" object:nil];

}


- (void)didTapDismissButton { [self dismissVC]; }

@end

File: Sources/Controllers/Misc/Views/ModalSheetVC.swift
import PhotosUI
import UIKit


protocol ModalSheetVCDelegate: AnyObject {
	func shouldReloadData(in modalSheetVC: ModalSheetVC)
}

/// Controller that'll show the modal sheet view
final class ModalSheetVC: UIViewController {

	let dataSource = ModalSheetDataSource()

	var headerView: NewIssuerOptionsHeaderView { return modalChildView.headerView }
	var childTableView: UITableView { return modalChildView.tableView }

	private let newIssuerVC = NewIssuerVC()
	private let toastView = ToastView()

	private var modalChildView: ModalChildView!
	private var navVC: UINavigationController!

	weak var delegate: ModalSheetVCDelegate?

	// ! Lifecycle

	required init?(coder: NSCoder) {
		super.init(coder: coder)
	}

	override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
		super.init(nibName: nil, bundle: nil)

		modalChildView = .init(dataSource: dataSource)
		modalChildView.delegate = self
		newIssuerVC.delegate = self

		view.backgroundColor = .clear
	}

	override func loadView() { view = modalChildView }

	override func viewWillAppear(_ animated: Bool) {
		super.viewWillAppear(animated)
		modalChildView.animateViews()
	}

	// ! Reusable funcs

	private func configureVC(
		_ vc: UIViewController,
		withTitle title: String,
		withItemImage image: UIImage,
		forSelector selector: Selector,
		isLeftBarButtonItem: Bool
	) {
		navVC = UINavigationController(rootViewController: vc)
		vc.title = title
		if isLeftBarButtonItem {
			vc.navigationItem.leftBarButtonItem = .getBarButtomItem(withImage: image, target: self, selector: selector)
		}
		else {
			vc.navigationItem.rightBarButtonItem = .getBarButtomItem(withImage: image, target: self, selector: selector)
		}
		navVC.modalTransitionStyle = .crossDissolve
		navVC.modalPresentationStyle = .fullScreen
	}

	private func dismissVC() {
		dismiss(animated: true, completion: nil)
		DispatchQueue.main.asyncAfter(deadline: .now() + 0.8) {
			self.modalChildView.animateDismiss { [weak self] _ in
				self?.dismiss(animated: true)
			}
		}
	}

	// ! Public

	/// Function to reload the child table view's data
	func reloadData() {
		modalChildView.reloadData()
	}

	/// Function to dismiss the current view controller being presented
	func shouldDismissVC() {
		modalChildView.animateDismiss { [weak self] _ in
			self?.dismiss(animated: true)
		}
	}

	/// Function to setup the backup options data source
	/// - Parameters:
	///		- buttonTarget: The button's target
	///		- selectors: An array of selectors for the buttons
	func setupBackupOptionsDataSource(buttonTarget target: Any?, selectors: [Selector]) {
		dataSource.setupBackupOptionsDataSource(buttonTarget: target, selectors: selectors)
	}

	/// Function to setup the make backup options data source
	/// - Parameters:
	///		- buttonTarget: The button's target
	///		- selectors: An array of selectors for the buttons
	func setupMakeBackupOptionsDataSource(buttonTarget target: Any?, selectors: [Selector]) {
		dataSource.setupMakeBackupOptionsDataSource(buttonTarget: target, selectors: selectors)
	}

}

// ! ModalChildViewDelegate

extension ModalSheetVC: ModalChildViewDelegate {

	@objc func didTapScanQRCodeButton(in modalChildView: ModalChildView) {
		let qrCodeVC = QRCodeVC()
		qrCodeVC.delegate = self

		configureVC(
			qrCodeVC,
			withTitle: "Scan QR Code",
			withItemImage: UIImage(systemName: "xmark.circle.fill") ?? UIImage(),
			forSelector: #selector(didTapDismissButton),
			isLeftBarButtonItem: true
		)
		present(navVC, animated: true)
	}

	@objc func didTapImportQRImageButton(in modalChildView: ModalChildView) {
		var configuration = PHPickerConfiguration()
		configuration.filter = PHPickerFilter.images

		let phPickerVC = PHPickerViewController(configuration: configuration)
		phPickerVC.delegate = self
		present(phPickerVC, animated: true)

		keyWindow.addSubview(toastView)
		keyWindow.pinToastToTheBottomCenteredOnTheXAxis(toastView, bottomConstant: -5)
	}

	@objc func didTapEnterManuallyButton(in modalChildView: ModalChildView) {
		configureVC(
			newIssuerVC,
			withTitle: "Enter QR Code",
			withItemImage: UIImage(systemName: "checkmark.circle.fill") ?? UIImage(),
			forSelector: #selector(didTapComposeButton),
			isLeftBarButtonItem: false
		)
		newIssuerVC.navigationItem.leftBarButtonItem = .getBarButtomItem(
			withImage: UIImage(systemName: "xmark.circle.fill") ?? UIImage(),
			target: self,
			selector: #selector(didTapDismissButton)
		)
		present(navVC, animated: true)
	}

	func didTapDimmedView(in modalChildView: ModalChildView) {
		modalChildView.animateDismiss { [weak self] _ in
			self?.dismiss(animated: true)
		}
	}

	func modalChildView(
		_ modalChildView: ModalChildView,
		didPanWithGesture gesture: UIPanGestureRecognizer,
		modifyingConstraint constraint: NSLayoutConstraint
	) {
		let translation = gesture.translation(in: view)
		let newHeight = modalChildView.currentSheetHeight - translation.y

		switch gesture.state {
			case .changed:
				if newHeight < modalChildView.kDefaultHeight {
					constraint.constant = newHeight
					constraint.isActive = true

					modalChildView.calculateAlpha(basedOnTranslation: translation)
					view.layoutIfNeeded()
				}
			case .ended:
				if newHeight < modalChildView.kDismissableHeight {
					modalChildView.animateDismiss { [weak self] _ in
						self?.dismissVC()
					}
				}
				else if newHeight < modalChildView.kDefaultHeight {
					modalChildView.animateSheetHeight(modalChildView.kDefaultHeight)
				}
			default: break
		}
	}

	@objc private func didTapComposeButton() {
		NotificationCenter.default.post(name: .shouldSaveDataNotification, object: nil)
	}

	@objc private func didTapDismissButton() { dismissVC() }

}

// ! NewIssuerVCDelegate

extension ModalSheetVC: NewIssuerVCDelegate {

	func shouldDismissVC(in newIssuerVC: NewIssuerVC) {
		delegate?.shouldReloadData(in: self)
		dismissVC()
	}

	func shouldPushAlgorithmVC(in newIssuerVC: NewIssuerVC) {
		let algorithmVC = AlgorithmVC()
		algorithmVC.title = "Algorithm"
		navVC.pushViewController(algorithmVC, animated: true)
	}

}

// ! QRCodeVCDelegate

extension ModalSheetVC: QRCodeVCDelegate {

	func didCreateIssuerOutOfQRCode(in qrCodeVC: QRCodeVC) {
		delegate?.shouldReloadData(in: self)
		dismissVC()
	}

}

// ! PHPickerViewControllerDelegate

extension ModalSheetVC: PHPickerViewControllerDelegate {

	func picker(_ picker: PHPickerViewController, didFinishPicking results: [PHPickerResult]) {
		guard !results.isEmpty else {
			dismissVC()
			return
		}
		results.first?.itemProvider.loadObject(ofClass: UIImage.self) { imageObject, error in
			guard let image = imageObject as? UIImage, error == nil else { return }

			let ciImage = CIImage(image: image) ?? CIImage()
			let options = [CIDetectorAccuracy: CIDetectorAccuracyHigh]
			let detector = CIDetector(ofType: CIDetectorTypeQRCode, context: nil, options: options)
			let features = detector?.features(in: ciImage) as? [CIQRCodeFeature] ?? []

			DispatchQueue.main.async {
				guard let otPauthString = features.first?.messageString else {
					self.toastView.fadeInOutToastView(withMessage: "No QR code was detected on this image.", finalDelay: 1.5)
					return
				}
				IssuerManager.sharedInstance.createIssuer(outOfOtPauthString: otPauthString) { isDuplicateItem, issuer in
					guard !isDuplicateItem else {
						self.toastView.fadeInOutToastView(withMessage: "Item already exists, updating it now.", finalDelay: 1.5)
						return
					}

					IssuerManager.sharedInstance.issuers.append(issuer)

					self.delegate?.shouldReloadData(in: self)
					self.dismissVC()
				}
			}
		}
	}

}
